Add ERC: Atomic ERC-7579 Module Install and Execute (#1253)
ethereum/ERCs: ERCS/erc-8043.md

---
eip: 8043

> title: Enable Mode for module installation in ERC-7579 Smart Accounts

Author: ```suggestion
title: Enable Mode for modules in ERC-7579 accounts
```

---
eip: 8043

> title: Enable Mode for modules in ERC-7579 accounts

Reviewer: The title of the proposal doesn't make it immediately obvious what the proposal does.

The description clears that up, but it feels like your burying the lede.

[ERC-7579]: ./eip-7579.md
[EIP-712]: ./eip-712.md
[ERC-1271]: ./eip-1271.md
[ERC-4337]: ./eip-4337.md
[ERC-7579]: ./eip-7579.md

## Motivation

Current [ERC-7579] implementations require separate transactions for module installation and usage, which can hinder user experience. Use cases such as session key and spending limit modules would benefit from the ability to install and use modules in a single operation. This standard allows applications to opt in to install and use multiple modules under a single signature and user operation, enabling greater flexibility—particularly for scenarios where multiple modules need to work together atomically.

This standard addresses several key issues:

- **User Experience**: Reduces transaction count from N+1 to 1 for installing N modules and using them
- **Consistency across implementations**: Provides a standard pattern that works across different [ERC-7579] account implementations
- **Security**: Maintains strong authorization through [EIP-712] signatures from account owners

> Real-world usage in production systems like Kernel has validated the need and feasibility of this pattern, particularly for session key management and dynamic module installation.

Reviewer: ```suggestion
Real-world usage in production systems has validated the need and feasibility of this pattern, particularly for session key management and dynamic module installation.
```

Best not to mention specific products.

### EIP-712 Message Structure

The message structure MUST be:

```solidity
bytes32 constant ENABLE_MODULES_TYPEHASH = keccak256(
    "EnableModules(uint256[] moduleTypeIds,address[] modules,bytes[] initDatas,uint256 nonce)"
);
```

### Validation Flow

Smart accounts implementing this standard on top of [ERC-4337]'s `validateUserOp` function:

1. MUST detect the enable mode prefix (`0x01`) in user operation nonces
2. MUST decode the signature according to the specified format
3. MUST validate the EIP-712 `installationSignature` using `isValidSignature`
4. SHOULD install each module using the standard ERC-7579 `installModule` flow using the corresponding `moduleTypeIds[i]`, `modules[i]`, and `initDatas[i]`
5. MUST continue with normal user operation validation using the extracted `userOpSignature`
6. MUST return `SIG_VALIDATION_FAILED` (`1`) if signature decoding fails, enabling signature is invalid, or any module installation fails

## Rationale

### Magic Nonce Prefix Choice

The prefix `0x01` was chosen because:

- It's unlikely to conflict with existing nonce usage patterns
- It's easily detectable with bitwise operations

> - It follows the pattern established by successful implementations like Kernel

Reviewer: ```suggestion
- It follows the pattern established by successful implementations
```
